### PR TITLE
Add UI for mailboxes, folders and messages (task #8830)

### PR DESCRIPTION
**Pull request by stepanov**

### CODE REVIEW

File: config/Migrations/20190303221533_AddIconToFolder.php [hunk 1]
```diff
@@ -0,0 +1,23 @@
+<?php
+use Migrations\AbstractMigration;
+
+class AddIconToFolder extends AbstractMigration
```
[the remark this answers is not in the document]
Comment (stepanov) [reply config/Migrations/20190303221533_AddIconToFolder.php]: @softius can we merge this PR? I need it in other branch to finish UI

File: src/Controller/MessagesController.php [hunk 2]
```diff
@@ -55,57 +65,77 @@ public function view(string $id = null)
...
+        $mailbox = $this->getMailbox($mailboxId);
         $message = $this->Messages->newEntity();
         if ($this->request->is('post')) {
             $data = $this->request->getData();
+            Assert::isArray($data);
+
             $data['from_user'] = $this->Auth->user('id');
             $data['status'] = $this->Messages->getNewStatus();
             $data['date_sent'] = $this->Messages->getDateSent();
+            $data['folder_id'] = $this->getFolderByName('Sent', $mailboxId);
```
Comment (softius) [inline src/Controller/MessagesController.php]: @stepanov I believe we can use `MailboxesTable::FOLDER_SENT` here

File: src/Template/Mailboxes/view.ctp [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+                                ]); ?>
+                            </div>
+                        </div>
+                    </div>
+                    <?php if (0 < $messages->count()) : ?>
+                    <div class="table-responsive mailbox-messages">
+                        <table id="folder-table" class="table table-hover table-striped">
+                            <thead>
+                                <th></th>
+                                <th><?= 'Sent' === $folderName ? __('To') : __('From') ?></th>
```
Comment (softius) [inline src/Template/Mailboxes/view.ctp]: @stepanov I believe we can use `MailboxesTable::FOLDER_SENT` here

File: src/Template/Mailboxes/view.ctp [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+                                <th><?= __('Subject') ?></th>
+                                <th><?= __('Date') ?></th>
+                            </thead>
+                            <tbody>
+                            <?php foreach ($messages as $message) : ?>
+                                <?php
+                                $messageUser = 'sent' === $folderName ? 'toUser' : 'fromUser';
```
Comment (softius) [inline src/Template/Mailboxes/view.ctp]: @stepanov same here

File: src/Template/Mailboxes/view.ctp [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+                                $messageUrl = $this->Url->build([
+                                    'plugin' => 'MessagingCenter',
+                                    'controller' => 'Messages',
+                                    'action' => 'view',
+                                    $message->id
+                                ]);
+                                ?>
+                                <tr>
+                                    <td class="mailbox-read text-center">
+                                        <?php if ('new' === $message->status && 'Sent' !== $folderName) : ?>
```
Comment (softius) [inline src/Template/Mailboxes/view.ctp]: @stepanov same here

File: src/Model/Table/MailboxesTable.php [hunk 6]
```diff
@@ -180,4 +191,51 @@ public function createDefaultMailbox(array $user) : EntityInterface
...
+     */
+    public function getInboxFolder(EntityInterface $mailbox) : string
+    {
+        if (empty($mailbox->get('folders'))) {
+            throw new InvalidArgumentException('No folder is created for that mailbox.');
+        }
+
+        $inboxFolderId = null;
+        foreach ($mailbox->get('folders') as $folder) {
+            if ($folder->get('name') === 'Inbox') {
```
Comment (softius) [inline src/Model/Table/MailboxesTable.php]: @stepanov @stepanov I believe we can use `MailboxesTable::FOLDER_INBOX` here

File: src/Template/Element/folders_list.ctp [hunk 7]
```diff
@@ -43,21 +25,26 @@ if (!isset($folder)) {
...
-            <li class="<?= $action === $folder ? ' active' : ''; ?>">
+        <?php foreach ($mailbox->get('folders') as $folder) : ?>
+            <li class="<?= $folder->get('name') === $folderName ? ' active' : ''; ?>">
             <?php
-            if ('inbox' === $action) {
-                $unreadCount = (int)$this->cell('MessagingCenter.Inbox::unreadCount', ['{{text}}'])->render();
-                if (0 < $unreadCount) {
-                    $options['label'] .= ' <span class="label label-primary pull-right">' . $unreadCount . '</span>';
+                $label = $folder->get('name');
+                if ('Inbox' === $folder->get('name')) {
```
Comment (softius) [inline src/Template/Element/folders_list.ctp]: @stepanov I believe we can use MailboxesTable::FOLDER_INBOX here

File: src/Template/Messages/view.ctp [hunk 8]
```diff
@@ -77,27 +70,30 @@ $restoreBtn = $this->Form->postLink(
...
+                            </span>
+                        </h5>
                     </div>
                     <div class="mailbox-controls with-border text-center">
                         <div class="btn-group">
                         <?php
-                        switch ($folder) {
-                            case 'inbox':
+                        switch ($folderName) {
+                            case 'Inbox':
```
Comment (softius) [inline src/Template/Messages/view.ctp]: @stepanov same here

File: src/Template/Messages/view.ctp [hunk 9]
```diff
@@ -106,28 +102,28 @@ $restoreBtn = $this->Form->postLink(
...
                 </div>
                 <div class="box-footer">
                     <div class="pull-right">
                         <?= $replyBtn; ?>
                     </div>
                     <?php
-                    switch ($folder) {
-                        case 'inbox':
+                    switch ($folderName) {
+                        case 'Inbox':
```
Comment (softius) [inline src/Template/Messages/view.ctp]: @stepanov same here

File: src/Shell/FetchMailShell.php [hunk 10]
```diff
@@ -84,21 +101,18 @@ protected function processMailbox(Mailbox $mailbox): void
         ];
 
         try {
-            $this->out('Fetching mail for [' . $mailbox->name . ']');
+            $this->out('Fetching mail for [' . $mailbox->get('name') . ']');
 
-            $settings = json_decode($mailbox->incoming_settings, true) ?? [];
+            $settings = json_decode($mailbox->get('incoming_settings'), true) ?? [];
```
Comment (georgeconstantinou) [inline src/Shell/FetchMailShell.php]: @stepanov we can get rid of JSON encoding/decoding by using JSON column type: https://book.cakephp.org/3.0/en/orm/saving-data.html#saving-complex-types

Comment (stepanov) [reply src/Shell/FetchMailShell.php]: @georgeconstantinou very good point, man

**Pull request merged**